_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+<script src="main.js"></script>
+</body>
+</html>
+```
+
+Add the `/profile` handler in the `HomeController` class:
+
+```java
+// src/main/java/com/example/demo/web/HomeController.java
+...
+
+@GetMapping("/profile")
+public String profile(@AuthenticationPrincipal OidcUser oidcUser, Model model) {
+    model.addAttribute("username", oidcUser.getEmail());
+    model.addAttribute("claims", oidcUser.getClaims());
+    return "profile";
+}
+
+...
+```
+
+### Configure Authentication Code Flow with PKCE at Auth0
+
+> Auth0 supports [OAuth 2.0 Security Best Current Practice](https://oauth.net/2/oauth-best-practice/), and you can register the application as a Regular Web Application (confidential client), and the Auth0 authorization server will honor the PKCE flow.
+
+Sign up at [Auth0](https://auth0.com/signup) and install the [Auth0 CLI](https://github.com/auth0/auth0-cli). Then in the command line run:
+
+```shell
+auth0 login
+```
+
+The command output will display a device confirmation code and open a browser session to activate the device.
+
+Using the Auth0 CLI, register the web application as an authentication client for Auht0:
+
+```shell
+auth0 apps create \
+  --name "Spring MVC" \
+  --description "Spring Boot Webapp" \
+  --type regular \
+  --callbacks http://localhost:4040/login/oauth2/code/okta \
+  --logout-urls http://localhost:4040 \
+  --reveal-secrets
+```
+
+Notice the output contains your Auth0 domain, clientId, and clientSecret.
+
+Rename `application.properties` to `application.yml` and add the following values:
+
+```yml
+# src/main/resources/application.yml
+server:
+  port: ${PORT}
+
+okta:
+  oauth2:
+    issuer: ${OKTA_OAUTH2_ISSUER}
+    client-id: ${OKTA_OAUTH2_CLIENT_ID}
+    client-secret: ${OKTA_OAUTH2_CLIENT_SECRET}
+
+logging:
+  level:
+    org.springframework.security: DEBUG
+    org.springframework.web: DEBUG
+```
+
+The Okta Spring Boot Starter will detect the presence of the properties above and auto-configure the Spring Security filter chain for OpenID Connect authentication. The configuration also enables security and web logs for analyzing the authentication flow.
+
+Create a `.env` file under the root project directory:
+
+```shell
+touch .env
+```
+
+Populate the `.env` with the following environment variables and values from the previous `auth0 apps create` command output:
+
+```shell
+PORT=4040
+OKTA_OAUTH2_ISSUER=https://<your-auth0-domain>/
+OKTA_OAUTH2_CLIENT_ID=<client-id>
+OKTA_OAUTH2_CLIENT_SECRET=<client-secret>
+```
+
+> The default client authentication method (how the application identifies itself when calling the token endpoint) in Auth0, when you register a regular web application, is `client_secret_post`, but it supports `client_secret_basic` as well. The default client authentication method in Spring Security for confidential clients is `client_secret_basic` if the provider supports it (available in the provider configuration metadata).
```

Comment:
does it require any config change anywhere?

Reply:
Added comment, no change required.